File: scm-java/src/main/scheme/TextStream.java
package scheme;

import java.io.PushbackReader;
import java.io.IOException;

public class TextStream {
    private PushbackReader reader;
    private String filename;
    private int line = 1;
    private int column = 0;

    public boolean foldCase = false;
    public boolean isOpen = true;

    public TextStream(PushbackReader reader, String filename) {
        this.reader = reader;
        this.filename = filename;
    }

    public void close() throws IOException {
        isOpen = false;
        this.reader.close();
    }

    public void setPosition(int line, int column) {
        this.line = line;
        this.column = column;
    }

    public String readToEnd() throws IOException {
        StringBuilder result = new StringBuilder();
        char[] buffer = new char[10240];
        int read = this.reader.read(buffer);
        while (read > 0) {
            result.append(buffer, 0, read);
            read = this.reader.read(buffer);
        }
        // TODO do we need to track the position here? probably not...
        return result.toString();
    }

    public String readLine() throws IOException {
        StringBuilder result = new StringBuilder();
        while (true) {
            int c = reader.read();
            if (c == -1) break;
            if (c == '\r' || c == '\n') {
                if (c == '\r' && peek() == '\n') {
                    reader.read();
                }
                return result.toString();
            }
            result.append((char) c);
        }
        if (result.length() > 0) {
            this.line++;
            this.column = 0;
            return result.toString();
        }
        return null;
    }

    public int read() throws IOException {
        int c = this.reader.read();
        if (c == '\n') {
            this.line++;
            this.column = 0;
        } else {
            this.column++;
        }
        return c;
    }

    public int peek() throws IOException {
        int c = this.reader.read();
        if (c != -1) {
            this.reader.unread(c);
        }
        return c;
    }

    public String filename() {
        return filename;
    }

    public int line() {
        return line;
    }

    public int column() {
        return column;
    }

    public SourcePos pos() {
        return new SourcePos(filename, line, column);
    }
}

File: scm-csharp/scheme/TextStream.cs
namespace scheme;

public class TextStream : IDisposable
{
    private TextReader reader;
    private string filename;
    private int line = 1;
    private int column = 0;

    public bool FoldCase { get; set; } = false;
    public bool IsOpen { get; private set; } = true;

    public TextStream(TextReader reader, string filename)
    {
	this.reader = reader;
	this.filename = filename;
    }

    public void Dispose()
    {
	this.reader.Dispose();
    }

    public void SetPosition(int line, int column)
    {
	this.line = line;
	this.column = column;
    }

    public string? ReadToEnd()
    {
	// TODO do we need to track the position here? probably not...
	return this.reader.ReadToEnd();
    }

    public string? ReadLine()
    {
        var result = this.reader.ReadLine();
        this.line++;
        this.column = 0;
        return result;
    }
    
    public int Read()
    {
	int c = this.reader.Read();
	if (c == '\n')
	{
	    this.line++;
	    this.column = 0;
	}
	else
	{
	    this.column++;
	}
	return c;
    }

    public int Peek()
    {
	return this.reader.Peek();
    }

    public void Close()
    {
        IsOpen = false;
	this.reader.Close();
    }

    public string Filename()
    {
	return filename;
    }
    
    public int Line()
    {
	return line;
    }

    public int Column()
    {
	return column;
    }

    public SourcePos Pos()
    {
	return new SourcePos(filename, line, column);
    }
}
